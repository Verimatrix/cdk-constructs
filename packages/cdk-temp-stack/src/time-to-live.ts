import * as path from 'path';
import { Rule, Schedule, RuleTargetInput } from '@aws-cdk/aws-events';
import { LambdaFunction } from '@aws-cdk/aws-events-targets';
import { Effect, PolicyStatement } from '@aws-cdk/aws-iam';
import { Code, Runtime, Function } from '@aws-cdk/aws-lambda';
import { Annotations, Aws, Construct, Duration, Stack } from '@aws-cdk/core';

export interface TimeToLiveProps {
  /**
   * Specifies the Time to Live (TTL) settings for the stack.
   */
  readonly ttl: Duration;
}

export class TimeToLive extends Construct {
  constructor(scope: Construct, id: string, props: TimeToLiveProps) {
    super(scope, id);

    Annotations.of(this).addInfo(`Warning! The stack destroys itself in ${props.ttl.toMinutes()} minutes.\n`);

    const deleteStack = new Function(this, 'DeleteStack', {
      runtime: Runtime.NODEJS_12_X,
      code: Code.fromAsset(path.join(__dirname, 'lambdas', 'delete-stack')),
      handler: 'index.handler',
    });

    deleteStack.addToRolePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['*'],
        resources: ['*'],
      }),
    );

    const rule = new Rule(this, 'TimeToLive', {
      schedule: Schedule.rate(props.ttl),
    });

    rule.addTarget(
      new LambdaFunction(deleteStack, {
        event: RuleTargetInput.fromObject({
          stackId: Aws.STACK_ID,
        }),
      }),
    );
  }

  protected onPrepare(): void {
    Stack.of(this).node.children.forEach((c) => {
      if (!this.node.findAll().includes(c)) {
        c.node.addDependency(this);
      }
    });
  }

  protected validate(): string[] {
    let count = 0;
    Stack.of(this).node.children.forEach((c) => {
      if (c instanceof TimeToLive) {
        count++;
      }
    });

    if (count > 1) {
      return [`Found ${count} instances of the TimeToLive construct in the stack. The construct may only be added once per stack.`];
    }

    return [];
  }
}
