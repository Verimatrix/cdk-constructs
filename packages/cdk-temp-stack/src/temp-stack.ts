import { Stack, StackProps, Duration } from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { TimeToLive } from './time-to-live';

export interface TempStackProps extends StackProps {
  /**
   * Specifies the Time to Live (TTL) settings for the stack.
   */
  readonly ttl: Duration;
}

export class TempStack extends Stack {
  constructor(scope: Construct, id: string, props: TempStackProps) {
    super(scope, id, props);

    new TimeToLive(this, 'TimeToLive', {
      ttl: props.ttl,
    });
  }
}
