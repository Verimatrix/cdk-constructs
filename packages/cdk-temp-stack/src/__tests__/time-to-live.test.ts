//import { SynthUtils } from '@aws-cdk/assert';
import { App, Duration, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import 'jest-cdk-snapshot';

import { TempStack } from '../temp-stack';
import { TimeToLive } from '../time-to-live';

test('default setup', () => {
  // GIVEN
  const stack = new Stack();

  // WHEN
  new TimeToLive(stack, 'TimeToLive', {
    ttl: Duration.minutes(10),
  });

  // THEN
  expect(stack).toMatchCdkSnapshot();
});

test('multiple instances', () => {
  // GIVEN
  const stack = new Stack();

  // WHEN
  new TimeToLive(stack, 'TimeToLive1', {
    ttl: Duration.minutes(10),
  });

  new TimeToLive(stack, 'TimeToLive2', {
    ttl: Duration.minutes(10),
  });

  // THEN
  expect(() => Template.fromStack(stack)).toThrowError(/Found 2 instances of the TimeToLive construct/);
});

test('one instance + temp stack', () => {
  // GIVEN
  const app = new App();

  // WHEN
  const stack = new TempStack(app, 'TestStack', {
    ttl: Duration.minutes(10),
  });

  new TimeToLive(stack, 'TimeToLive2', {
    ttl: Duration.minutes(10),
  });

  // THEN
  expect(() => Template.fromStack(stack)).toThrowError(/Found 2 instances of the TimeToLive construct/);
});
