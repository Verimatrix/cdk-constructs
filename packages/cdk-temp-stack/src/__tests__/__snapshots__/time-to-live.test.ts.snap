// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default setup 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TimeToLive7895C003": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(10 minutes)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "TimeToLiveDeleteStack08274E9B",
                "Arn",
              ],
            },
            "Id": "Target0",
            "Input": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{\\"stackId\\":\\"",
                  Object {
                    "Ref": "AWS::StackId",
                  },
                  "\\"}",
                ],
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "TimeToLiveAllowEventRuleTimeToLiveDeleteStack377107EF4EAD2805": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "TimeToLiveDeleteStack08274E9B",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "TimeToLive7895C003",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TimeToLiveDeleteStack08274E9B": Object {
      "DependsOn": Array [
        "TimeToLiveDeleteStackServiceRoleDefaultPolicy08F3148E",
        "TimeToLiveDeleteStackServiceRole5597A6E0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "cecce40e09824e72938e9e845377beb0cec84535601ffd985a78c42195247735.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TimeToLiveDeleteStackServiceRole5597A6E0",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TimeToLiveDeleteStackServiceRole5597A6E0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TimeToLiveDeleteStackServiceRoleDefaultPolicy08F3148E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TimeToLiveDeleteStackServiceRoleDefaultPolicy08F3148E",
        "Roles": Array [
          Object {
            "Ref": "TimeToLiveDeleteStackServiceRole5597A6E0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
